File: backend/app/schemas/position_workbench.py
from __future__ import annotations

from datetime import date, datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, EmailStr

from app.core.enums import PositionAssignmentType, ReportingLineType, UserStatus
from app.schemas.profiles import PositionRead


class PositionImpactCountsRead(BaseModel):
  assignee_count: int
  primary_count: int
  secondary_count: int
  department_count: int
  active_reporting_line_count: int


class PositionWorkbenchSummaryRead(BaseModel):
  total_positions: int
  active_positions: int
  total_assignees: int


class PositionWorkbenchCatalogItemRead(BaseModel):
  model_config = ConfigDict(from_attributes=True)

  id: UUID
  code: str
  name: str
  level: str | None
  extra_metadata: dict[str, Any]
  is_active: bool
  created_at: datetime
  updated_at: datetime
  impact: PositionImpactCountsRead


class PositionWorkbenchCatalogRead(BaseModel):
  summary: PositionWorkbenchSummaryRead
  positions: list[PositionWorkbenchCatalogItemRead]
  as_of: date


class PositionAssigneeUserSummaryRead(BaseModel):
  user_id: UUID
  email: EmailStr
  real_name: str | None
  employee_no: str | None
  status: UserStatus
  department_id: UUID | None
  department_name: str | None


class PositionAssignmentDetailRead(BaseModel):
  id: UUID
  user: PositionAssigneeUserSummaryRead
  department_id: UUID
  department_name: str | None
  assignment_type: PositionAssignmentType
  is_primary: bool
  starts_at: date
  ends_at: date | None
  is_effective: bool


class PositionReportingLineDetailRead(BaseModel):
  id: UUID
  user: PositionAssigneeUserSummaryRead
  manager_user_id: UUID
  manager_label: str | None
  department_id: UUID | None
  department_name: str | None
  line_type: ReportingLineType
  is_primary: bool
  starts_at: date
  ends_at: date | None
  is_effective: bool


class PositionEffectiveRangeRead(BaseModel):
  earliest_starts_at: date | None
  latest_ends_at: date | None
  open_ended_assignment_count: int


class PositionWorkbenchDetailRead(BaseModel):
  position: PositionRead
  impact: PositionImpactCountsRead
  assignments: list[PositionAssignmentDetailRead]
  reporting_lines: list[PositionReportingLineDetailRead]
  effective_range: PositionEffectiveRangeRead
  as_of: date

File: frontend/src/api/position-workbench.ts
import type {
  Position,
  PositionAssignmentType,
  ReportingLineType,
  UserStatus,
} from '@/types/api'
import { http } from './http'

export interface PositionImpactCounts {
  assignee_count: number
  primary_count: number
  secondary_count: number
  department_count: number
  active_reporting_line_count: number
}

export interface PositionWorkbenchCatalogItem extends Position {
  impact: PositionImpactCounts
}

export interface PositionWorkbenchCatalog {
  summary: {
    total_positions: number
    active_positions: number
    total_assignees: number
  }
  positions: PositionWorkbenchCatalogItem[]
  as_of: string
}

export interface PositionAssigneeUserSummary {
  user_id: string
  email: string
  real_name: string | null
  employee_no: string | null
  status: UserStatus
  department_id: string | null
  department_name: string | null
}

export interface PositionAssignmentDetail {
  id: string
  user: PositionAssigneeUserSummary
  department_id: string
  department_name: string | null
  assignment_type: PositionAssignmentType
  is_primary: boolean
  starts_at: string
  ends_at: string | null
  is_effective: boolean
}

export interface PositionReportingLineDetail {
  id: string
  user: PositionAssigneeUserSummary
  manager_user_id: string
  manager_label: string | null
  department_id: string | null
  department_name: string | null
  line_type: ReportingLineType
  is_primary: boolean
  starts_at: string
  ends_at: string | null
  is_effective: boolean
}

export interface PositionWorkbenchDetail {
  position: Position
  impact: PositionImpactCounts
  assignments: PositionAssignmentDetail[]
  reporting_lines: PositionReportingLineDetail[]
  effective_range: {
    earliest_starts_at: string | null
    latest_ends_at: string | null
    open_ended_assignment_count: number
  }
  as_of: string
}

export async function getPositionWorkbenchCatalog(asOf?: string): Promise<PositionWorkbenchCatalog> {
  const { data } = await http.get<PositionWorkbenchCatalog>('/positions/workbench', {
    params: asOf ? { as_of: asOf } : undefined,
  })
  return data
}

export async function getPositionWorkbenchDetail(
  positionId: string,
  asOf?: string,
): Promise<PositionWorkbenchDetail> {
  const { data } = await http.get<PositionWorkbenchDetail>(`/positions/${positionId}/workbench`, {
    params: asOf ? { as_of: asOf } : undefined,
  })
  return data
}
